cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -97,8 +97,8 @@ type ClusterConfiguration struct {
 	// Scheduler contains extra settings for the scheduler control plane component
 	Scheduler ControlPlaneComponent
 
-	// DNS defines the options for the DNS add-on installed in the cluster.
-	DNS DNS
+	// AddOns defines a map of addons deployed in this cluster. The key is the addon kind.
+	AddOns map[string]AddOn
```

Comment:
While it is clear to me that this enable user to customize the list of addons, I'm still missing if/how this will be reflected on how the user are expected to pass addons configurations.
Today this is handled inconsistently in kubeadm:
- kube-proxy via component config
- CoreDNS & kube-DNS not supported
I would like to see this somehow harmonized and made consistent, but TBH I don't understand if this is under the kubeadm responsibility or under the addon installer responsibility

Reply:
@rosti should confirm, but i'm seeing this feature as a stepping stone before kubeadm transition into using the addon installer types eventually a this getting deprecation.

but if that's the case this is weird because with each iteration of our API which should be approaching stability. Removing the list of addons later on and delegating it to an addon installer API will probably break users.

allowing addon manifest customization goes back to my proposal from 2018, which was rejected due to upgrade complications.

Reply:
My current idea about this when the cluster addon installer is ready is the following:
```yaml
apiVersion: kubeadm.k8s.io/v1beta4
kind: ClusterConfiguration
addons:
- kind: KubeProxy  # This is a complete GVK, forwarded as is to the cluster addons installer
  apiVersion: addons.config.k8s.io
  ...
---
# The component config (as is done currently by kubeadm), forwarded to the cluster addons installer as a component config
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
...
```
It might not be ideal though. We need more discussion around this.